mlrun/k8s_utils.py

```diff
@@ -228,3 +230,322 @@ def handle_invalid(message):
+            enriched_node_selector,
+            enriched_tolerations,
+            enriched_affinity,
+            preemptible_tolerations,
+        )
+
+    return (
+        enriched_node_selector,
+        enriched_tolerations,
+        _prune_empty_affinity(enriched_affinity),
+    )
+
+
+def _handle_prevent_mode(node_selector, tolerations, affinity, preemptible_tolerations):
+    def _prune_tolerations(to_remove: list[kubernetes.client.V1Toleration]):
+        return [t for t in tolerations if t not in to_remove]
+
+    # Ensure no preemptible node tolerations
+    tolerations = _prune_tolerations(preemptible_tolerations)
+
+    # Purge affinity preemption-related configuration
+    affinity = _prune_affinity_node_selector_requirement(
+        generate_preemptible_node_selector_requirements(
+            mlrun.common.schemas.NodeSelectorOperator.node_selector_op_in.value
+        ),
+        affinity=affinity,
+    )
+
+    # Remove preemptible nodes constraint
+    node_selector = _prune_node_selector(
+        mlconfig.get_preemptible_node_selector(),
+        enriched_node_selector=node_selector,
+    )
+
+    # Use anti-affinity only if no tolerations configured
+    if not preemptible_tolerations:
+        affinity = _override_required_during_scheduling_ignored_during_execution(
+            kubernetes.client.V1NodeSelector(
+                node_selector_terms=generate_preemptible_nodes_anti_affinity_terms()
+            ),
+            affinity,
+        )
+
+    return node_selector, tolerations, affinity
+
+
+def _handle_constrain_mode(
+    node_selector, tolerations, affinity, preemptible_tolerations
+):
+    tolerations = _merge_tolerations(tolerations, preemptible_tolerations)
+
+    affinity = _override_required_during_scheduling_ignored_during_execution(
+        kubernetes.client.V1NodeSelector(
+            node_selector_terms=generate_preemptible_nodes_affinity_terms()
+        ),
+        affinity=affinity,
+    )
+
+    return node_selector, tolerations, affinity
+
+
+def _handle_allow_mode(node_selector, tolerations, affinity, preemptible_tolerations):
+    for op in [
+        mlrun.common.schemas.NodeSelectorOperator.node_selector_op_not_in.value,
+        mlrun.common.schemas.NodeSelectorOperator.node_selector_op_in.value,
+    ]:
+        affinity = _prune_affinity_node_selector_requirement(
+            generate_preemptible_node_selector_requirements(op),
+            affinity=affinity,
+        )
+
+    node_selector = _prune_node_selector(
+        mlconfig.get_preemptible_node_selector(),
+        enriched_node_selector=node_selector,
+    )
+
+    tolerations = _merge_tolerations(tolerations, preemptible_tolerations)
+    return node_selector, tolerations, affinity
+
+
+def _merge_tolerations(
+    existing: list[kubernetes.client.V1Toleration],
+    to_add: list[kubernetes.client.V1Toleration],
+) -> list[kubernetes.client.V1Toleration]:
+    merged = existing.copy()
```

Comment:
from looking at your uses of this method, it seems inefficient to copy and return the `existing` list. 
I'd just merge the tolerations from `to_add` into the original `existing` as a reference and no need to return (even though you can return it)